contrib/babelfishpg_tsql/src/fts.c

```diff
@@ -64,3 +68,204 @@ babelfish_fts_rewrite(PG_FUNCTION_ARGS)
+    }
+
+    input_text = PG_GETARG_TEXT_P(0);
+    input_str = text_to_cstring(input_text);
+    trimmed_input_str = pstrdup(input_str);
+    
+    /*
+     * Remove leading and trailing whitespaces between single quotes. 
+     */
+    trim(trimmed_input_str, false);
+
+    input_length = strlen(trimmed_input_str);
+
+    if (input_length >= 2 && trimmed_input_str[0] == '"' && trimmed_input_str[input_length - 1] == '"') 
+    {
+        /* 
+         * Remove leading and trailing whitespaces between double quotes.
+         */
+        trim(trimmed_input_str, true);
+    }
+
+    input_length = strlen(trimmed_input_str);
+    
+    if(!trimmed_input_str || input_length == 0)
+    {
+        ereport(ERROR,
+            (errcode(ERRCODE_INTERNAL_ERROR),
+            errmsg("Null or empty full-text predicate.")));
+    }
+        
+    left_ptr = trimmed_input_str;
+    right_ptr = trimmed_input_str + input_length - 1;
+    while (left_ptr <= right_ptr)
+    {
+        if (strchr(specialChars, *left_ptr) != NULL) 
+        {
+            pfree(input_str);
+            resetStringInfo(&normalized_str);
+            ereport(ERROR,
+                (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+                 errmsg("Special characters in the freetext search string are not currently supported in Babelfish")));
+        }
+
+        /* 
+         * Removing multiple spaces, tabs, newlines and double quotes from within the search string
+         * If any of the space or quote is encountered, we remove all their next occurances
+         * before end of the input or if next word is encountered and
+         * we append a single space instead.
+         * Case 1: '"word1   word2"' = 'word1 word2'
+         * Case 2: '"word1  "  word2 "' = 'word1 word2'
+         * Case 3: '"word1' + CHAR(9) + '"' = 'word1'
+         * Case 3: '"word1' + CHAR(10) + '" word2' = 'word1 word2'
+         * Case 4: '"word1  ' + CHAR(10) + ' ' + CHAR(10) + '"" = 'word1'
+         * Case 5: '"word1  ' + CHAR(9) + ' ' + CHAR(10) + ' ""  word2   "' = 'word1 word2'
+         */
+        else if ((left_ptr <= right_ptr) && (*left_ptr == ' ' || *left_ptr == '\t' || *left_ptr == '\n' || *left_ptr == '"'))
+        {
+            while ((left_ptr <= right_ptr) && (*left_ptr == ' ' || *left_ptr == '\t' || *left_ptr == '\n' || *left_ptr == '"'))
+            {
+                left_ptr++;
+            }
+            
+            if(left_ptr <= right_ptr && strlen(normalized_str.data) != 0)
+            {
+                appendStringInfoChar(&normalized_str, ' ');
+            }
+        }
+
+        else
+        {
+            appendStringInfoChar(&normalized_str, *left_ptr);
+            left_ptr++;
+        }
+    }
+
+    trimmed_input_str = normalized_str.data;
+    
+    if (!pltsql_allow_fulltext_parser)
+    {
+        ereport(ERROR,
+            (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+            errmsg("Full Text Search is not yet supported.")));
+    }
```

Comment:
Please add this check at the starting, we don't want to perform any kind of processing if the feature is not even enabled.